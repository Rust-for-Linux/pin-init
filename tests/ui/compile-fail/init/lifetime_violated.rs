use pin_init::{init, InPlaceInit, Init};

#[derive(Debug)]
struct Foo {
    a: i32,
}

impl Foo {
    fn new(val: &i32) -> impl Init<Self> + use<'_> {
        init!(Self { a: *val })
    }
}

#[derive(Debug)]
struct Bar {
    foo: Foo,
}

impl Bar {
    fn new(foo: impl Init<Foo>) -> impl Init<Self> {
        init!(Self { foo <- foo })
    }
}

fn main() {
    // problematic:
    let foo;
    {
        let val = 42;
        foo = Foo::new(&val);
    }
    let bar = Box::init(Bar::new(foo));
    println!("{bar:?}");

    // okay:
    let val = 42;
    let foo = Foo::new(&val);
    let bar = Box::init(Bar::new(foo));
    println!("{bar:?}");
}
