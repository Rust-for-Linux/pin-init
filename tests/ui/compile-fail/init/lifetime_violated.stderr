error[E0597]: `val` does not live long enough
  --> tests/ui/compile-fail/init/lifetime_violated.rs:30:24
   |
29 |         let val = 42;
   |             --- binding `val` declared here
30 |         foo = Foo::new(&val);
   |                        ^^^^ borrowed value does not live long enough
31 |     }
   |     - `val` dropped here while still borrowed
32 |     let bar = Box::init(Bar::new(foo));
   |                                  --- borrow later used here
